File: PMF_m/run_tp.m
clc; close all; clear;

addpath('src');
logger('===============================');
logger('PMF: Probabilistic Matrix Factorization.');
logger('===============================');

% * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
% config area 
dataPath = '../data/dataset#1/rawData/tpMatrix.txt';
dataFolder = '../data/dataset#1/generatedData/TP/density_';
outPath = 'result/tpResult_';
density = [0.01 : 0.01 : 0.05, 0.1 : 0.05 : 0.5]; % matrix density
rounds = 20; % how many runs are performed at each matrix density
dimension = 10; % dimenisionality of the latent factors
etaInit = 0.001; % inital learning rate. We use line search to find the best eta at each iteration
lambda = 800; % regularization parameter
maxIter = 500; % the max iterations
saveTimeInfo = false; % whether to keep track of the running time
debugMode = false; % whether to log some debugging info
% * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *

% set parameters of PMF algorithms
paraStruct = struct('rounds', rounds, 'dimension', dimension, 'etaInit', etaInit, 'lambda', lambda,... 
    'maxIter', maxIter, 'saveTimeInfo', saveTimeInfo, 'debugMode', debugMode); 

% create resultFolder
if exist('result', 'dir') == 0
    system('mkdir result');
end

% load the dataset
logger(sprintf('Load data: %s', dataPath));
dataMatrix = load(dataPath);

% run for each density
runTime = tic; % start timing
for i = 1 : length(density)
    PMF(dataMatrix, dataFolder, outPath, density(i), paraStruct);
end  

logger('===============================');
logger(sprintf('All done. Total running time: %f s', toc(runTime)));
logger('===============================');
rmpath('src');

 

File: PMF/run_tp.py
########################################################
# run_tp.py 
# Author: Jamie Zhu <jimzhu@GitHub>
# Created: 2014/2/6
# Last updated: 2014/5/4
# Implemented approach: PMF
# Evaluation metrics: MAE, NMAE, RMSE, MRE, NPRE
########################################################

import numpy as np
import os, sys, time
sys.path.append('src')
# Build external model
if not os.path.isfile('src/PMF.so'):
	print 'Lack of PMF.so. Please first build the cpp code into PMF.so: '
	print 'python setup.py build_ext --inplace'
	sys.exit()
from utilities import *
from predict import *
 

#########################################################
# config area
#
para = {'dataPath': '../data/dataset#1/tpMatrix.txt',
		'outPath': 'result/tpResult_',
		'metrics': ['MAE', 'NMAE', 'RMSE', 'MRE', 'NPRE'], # delete where appropriate
		# matrix density
		'density': list(np.arange(0.01, 0.06, 0.01)) 
					+ list(np.arange(0.10, 0.31, 0.05)), 
		'rounds': 20, # how many runs are performed at each matrix density
		'dimension': 10, # dimenisionality of the latent factors
		'etaInit': 0.001, # inital learning rate. We use line search
						 # to find the best eta at each iteration
		'lambda': 800, # regularization parameter
		'maxIter': 300, # the max iterations
		'saveTimeInfo': False, # whether to keep track of the running time
		'saveLog': False, # whether to save log into file
		'debugMode': False # whether to record the debug info
		}

initConfig(para)
#########################################################


startTime = time.clock() # start timing
logger.info('==============================================')
logger.info('PMF: Probabilistic Matrix Factorization.')
logger.info('Load data: %s'%para['dataPath'])
dataMatrix = np.loadtxt(para['dataPath']) 

# run for each density
for density in para['density']:
    predict(dataMatrix, density, para)

logger.info(time.strftime('All done. Total running time: %d-th day - %Hhour - %Mmin - %Ssec.',
         time.gmtime(time.clock() - startTime)))
logger.info('==============================================')
sys.path.remove('src')